aquadoggo/src/validation.rs

```diff
@@ -0,0 +1,335 @@
+}
+
+/// Verify that an entry's claimed log id matches what we expect from the claimed document id.
+///
+/// This method handles both the case where the claimed log id already exists for this author
+/// and where it is a new log.
+///
+/// The following steps are taken:
+/// - Retrieve the stored log id for the claimed document id
+///   - If found, ensure it matches the claimed log ig
+///   - If not found retrieve the next available log id for this author and ensure that matches
+pub async fn verify_log_id<S: StorageProvider>(
+    store: &S,
+    author: &Author,
+    claimed_log_id: &LogId,
+    document_id: &DocumentId,
+) -> Result<()> {
+    // Check if there is a log_id registered for this document and public key already in the store.
+    match store.get(author, document_id).await? {
+        Some(expected_log_id) => {
+            // If there is, check it matches the log id encoded in the entry.
+            ensure!(
+                *claimed_log_id == expected_log_id,
+                anyhow!(
+                    "Entry's claimed log id of {} does not match expected log id of {} for given author and document",
+                    claimed_log_id.as_u64(),
+                    expected_log_id.as_u64()
+                )
+            );
+        }
+        None => {
+            // If there isn't, check that the next log id for this author matches the one encoded in
+            // the entry.
+            let expected_log_id = next_log_id(store, author).await?;
+
+            ensure!(
+                *claimed_log_id == expected_log_id,
+                anyhow!(
+                    "Entry's claimed log id of {} does not match expected next log id of {} for given author",
+                    claimed_log_id.as_u64(),
+                    expected_log_id.as_u64()
+                )
+            );
+        }
+    };
+    Ok(())
+}
+
+// Get the _expected_ skiplink for the passed entry.
+//
+// This method retrieves the expected skiplink given the author, log and seq num
+// of an entry. It _does not_ verify that it matches the claimed skiplink
+// encoded on the passed entry.
+//
+// If the expected skiplink could not be found in the database an error is returned.
+pub async fn get_expected_skiplink<S: StorageProvider>(
+    store: &S,
+    author: &Author,
+    log_id: &LogId,
+    seq_num: &SeqNum,
+) -> Result<S::StorageEntry> {
+    ensure!(
+        !seq_num.is_first(),
+        anyhow!("Entry with seq num 1 can not have skiplink")
+    );
+    // Derive the expected skiplink seq number from this entries claimed sequence number
+    let expected_skiplink = match seq_num.skiplink_seq_num() {
+        // Retrieve the expected skiplink from the database
+        Some(seq_num) => {
+            let expected_skiplink = store.get_entry_at_seq_num(author, log_id, &seq_num).await?;
+            expected_skiplink
+        }
+        // Or if there is no skiplink for entries at this sequence number return None
+        None => None,
```

Comment:
Shouldn't the `ensure` from the first line of this method happen here instead?

Reply:
Yes, we could do it there :+1: